04.ls/ls.rb

```diff
@@ -20,6 +53,77 @@ def prepare_files(options)
   options['r'] ? sorted_files.reverse : sorted_files
 end
 
+def prepare_file_details(files)
+  file_details = []
+  blocks = 0
+
+  files.each do |filename|
+    filestat = File.stat(filename)
+    blocks += filestat.blocks
+    file_details << build_file_detail(filename, filestat)
+  end
+  [blocks, file_details]
+end
+
+def build_file_detail(filename, filestat)
+  file_permissions = build_file_permissions(filestat)
+  {
+    permission: file_permissions,
+    link: filestat.nlink,
+    owner_name: Etc.getpwuid(filestat.uid).name,
+    group_name: Etc.getgrgid(filestat.gid).name,
+    size: filestat.size,
+    timestamp: filestat.mtime.strftime('%b %e %H:%M'),
+    name: filename
+  }
+end
+
+def build_file_permissions(filestat)
+  octal_mode = filestat.mode.to_s(8)
+
+  file_permissions = [FILE_TYPE[filestat.ftype]]
+  special_permission_bits = octal_mode.slice(-4, 1)
+
+  octal_mode.chars[-3..].each_with_index do |bit, idx|
+    permission = PERMISSIONS[bit]
+    special_flag = SPECIAL_FLAGS[special_permission_bits]
+    file_permissions << if special_flag && special_flag[:idx] == idx
+                          update_permission(special_flag, permission)
+                        else
+                          permission.values_at(:r, :w, :x).join
+                        end
+  end
+
+  file_permissions.join
```

Comment:
```suggestion
  foo = FILE_TYPE[filestat.ftype]
  special_permission_bits = octal_mode.slice(-4, 1)

  bar = octal_mode.chars[-3..].map.with_index do |bit, idx|
    permission = PERMISSIONS[bit]
    special_flag = SPECIAL_FLAGS[special_permission_bits]
    if special_flag && special_flag[:idx] == idx
      update_permission(special_flag, permission)
    else
      permission.values_at(:r, :w, :x).join
    end
  end

  [foo, *bar].join
```

変数を二つに分けた方がコードがスッキリしそう

Reply:
ありがとうございます！変数を2つに分けました。